File: vespalib/src/vespa/vespalib/net/tls/maybe_tls_crypto_engine.cpp
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "maybe_tls_crypto_engine.h"
#include "maybe_tls_crypto_socket.h"
#include "crypto_codec.h"

namespace vespalib {

MaybeTlsCryptoEngine::~MaybeTlsCryptoEngine() = default;

std::unique_ptr<net::tls::CryptoCodec>
MaybeTlsCryptoEngine::create_tls_client_crypto_codec(const SocketHandle &socket, const SocketSpec &spec) {
    return _tls_engine->create_tls_client_crypto_codec(socket, spec);
}

std::unique_ptr<net::tls::CryptoCodec>
MaybeTlsCryptoEngine::create_tls_server_crypto_codec(const SocketHandle &socket) {
    return _tls_engine->create_tls_server_crypto_codec(socket);
}

CryptoSocket::UP
MaybeTlsCryptoEngine::create_client_crypto_socket(SocketHandle socket, const SocketSpec &spec)
{
    if (_use_tls_when_client) {
        return _tls_engine->create_client_crypto_socket(std::move(socket), spec);
    } else {
        return _null_engine->create_client_crypto_socket(std::move(socket), spec);
    }
}

CryptoSocket::UP
MaybeTlsCryptoEngine::create_server_crypto_socket(SocketHandle socket)
{
    return std::make_unique<MaybeTlsCryptoSocket>(std::move(socket), _tls_engine);
}

} // namespace vespalib

File: jrt/src/com/yahoo/jrt/MaybeTlsCryptoEngine.java
// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.jrt;

import java.nio.channels.SocketChannel;

/**
 * A crypto engine that supports both tls encrypted connections and
 * unencrypted connections. The use of tls for incoming connections is
 * auto-detected using clever heuristics. The use of tls for outgoing
 * connections is controlled by the useTlsWhenClient flag given to the
 * constructor.
 */
public class MaybeTlsCryptoEngine implements CryptoEngine {

    private final TlsCryptoEngine tlsEngine;
    private final boolean useTlsWhenClient;

    public MaybeTlsCryptoEngine(TlsCryptoEngine tlsEngine, boolean useTlsWhenClient) {
        this.tlsEngine = tlsEngine;
        this.useTlsWhenClient = useTlsWhenClient;
    }

    @Override
    public CryptoSocket createClientCryptoSocket(SocketChannel channel, Spec spec) {
        if (useTlsWhenClient) {
            return tlsEngine.createClientCryptoSocket(channel, spec);
        } else {
            return new NullCryptoSocket(channel, false);
        }
    }

    @Override
    public CryptoSocket createServerCryptoSocket(SocketChannel channel) {
        return new MaybeTlsCryptoSocket(channel, tlsEngine);
    }

    @Override
    public String toString() { return "MaybeTlsCryptoEngine(useTlsWhenClient:" + useTlsWhenClient + ")"; }

    @Override
    public void close() {
        tlsEngine.close();
    }
}
